neonvm/controllers/virtualmachine_controller.go

```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
-					return err
-				}
-			} else if !apierrors.IsNotFound(err) {
+		// Always delete runner pod. Otherwise, we could end up with one container succeeded/failed
+		// but the other one still running (meaning that the pod still ends up Running).
+		vmRunner := &corev1.Pod{}
+		err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
+		if err == nil {
+			// delete current runner
+			if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
 				return err
 			}
-			// do cleanup
+		} else if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		// We must keep the VM status the same until we know the neonvm-runner container has been
+		// terminated, otherwise we could end up starting a new runner pod while the VM in the old
+		// one is still running.
+		//
+		// Note that this is required because 'VmSucceeded' and 'VmFailed' are true if *at least
+		// one* container inside the runner pod has finished; the other may still be running.
+		if apierrors.IsNotFound(err) || runnerContainerStopped(vmRunner) {
+			// clean up status, but keep the phase; we'll use it below (and possibly keep it around to
+			// be visible to the user if we don't intend to restart).
+			phase := virtualmachine.Status.Phase
 			virtualmachine.Cleanup()
-		case vmv1.RestartPolicyOnFailure:
-			log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
-			// get runner to delete
-			found := true
-			vmRunner := &corev1.Pod{}
-			err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
-			if apierrors.IsNotFound(err) {
-				found = false
-			} else if err != nil {
-				return err
+			virtualmachine.Status.Phase = phase
+
+			shouldRestart := virtualmachine.Spec.RestartPolicy == vmv1.RestartPolicyAlways ||
+				(virtualmachine.Spec.RestartPolicy == vmv1.RestartPolicyOnFailure && virtualmachine.Status.Phase == vmv1.VmFailed)
+
+			if shouldRestart {
+				log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
+				virtualmachine.Status.Phase = "" // reset to trigger restart
 			}
-			// delete runner only when VM failed
-			if found && virtualmachine.Status.Phase == vmv1.VmFailed {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
+
+			// TODO for RestartPolicyNever: implement TTL or do nothing
+		}
+	default:
+		// do nothing
+	}
+
+	return nil
+}
+
+type runnerStatus string
+
+const (
+	runnerUnknown   runnerStatus = "Unknown"
+	runnerPending   runnerStatus = "Pending"
+	runnerRunning   runnerStatus = "Running"
+	runnerFailed    runnerStatus = "Failed"
+	runnerSucceeded runnerStatus = "Succeeded"
+)
+
+func runnerContainerStatus(pod *corev1.Pod) runnerStatus {
```

Comment:
Is there a way to declare in k8s if one of the containers in the pod fails, then the entire pod also should fail?

Reply:
Also this function doesn't return runner container status exactly, which is misleading.

Reply:
> Is there a way

I spent a while looking and couldn't find any way to do this. K8s 1.29 introduced ["sidecar containers"](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) as an explicit feature, which _do_ allow this.